xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+        """
+        self.add(callable)
+        return callable
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+
+        The caller may send() new positional arguments (eg, to implement
+        modifying semantics). Keyword arguments cannot be modified this way.
+        """
+        for handler in self:
```

Comment:
> I can't imagine a use case where the handler couldn't fail fast instead of having an extra lambda (or worse, a full function).

Git, hg, and other version control hooks.  It takes much longer and much more logic to figure out if you are in a repo than it does to execute whatever action you want to perform.

However, broadly speaking, validators help separate out the concern of "should I do x" from actually performing x.  You end up with two, smaller, better defined, and more testable functions  rather than one function that does everything.

Having optional validators is an essential feature for me.  If you don't wish to add them, that is fine, but I will probably put in a PR to add them.  Also it is only a few lines of code to implement, so it is not that big of a deal.  If devs don't want to use them , that is fine. That is why they are optional.